File: crates/solobase-cloudflare/worker/usage.ts
/**
 * Project usage tracking and plan limit enforcement.
 *
 * Tracks API request counts per project per month in D1.
 * Checks limits before dispatching to blocks.
 *
 * Table: project_usage
 *   - project_id TEXT
 *   - month TEXT (YYYY-MM)
 *   - requests INTEGER
 *   - r2_bytes INTEGER
 *   - addon_requests INTEGER (purchased add-on capacity)
 *   - addon_r2_bytes INTEGER
 *   - addon_d1_bytes INTEGER
 */

import type { ProjectConfig, PlanLimits } from './types';
import { getPlanLimits } from './types';

/**
 * Result of a usage check. Either:
 * - allowed (no error, optional warning)
 * - blocked (error message, request should be rejected)
 */
export interface UsageCheckResult {
  /** If set, the request should be rejected with 429. */
  error: string | null;
  /** If set, add as X-Solobase-Warning header (request still succeeds). */
  warning: string | null;
}

/**
 * Check if the project is within their plan limits and increment the request counter.
 */
export async function checkAndIncrementUsage(
  db: D1Database,
  project: ProjectConfig,
): Promise<UsageCheckResult> {
  const ok: UsageCheckResult = { error: null, warning: null };

  // Skip checks for platform
  if (project.plan === 'platform') return ok;

  const limits = getPlanLimits(project.plan);
  const month = currentMonth();

  // Check subscription status if project has an owner
  if (project.owner_user_id) {
    const sub = await db.prepare(
      'SELECT status, grace_period_end FROM subscriptions WHERE user_id = ? LIMIT 1'
    ).bind(project.owner_user_id).first<{ status: string; grace_period_end: string | null }>();

    if (sub) {
      if (sub.status === 'cancelled' || sub.status === 'suspended') {
        return { error: 'Subscription inactive. Please resubscribe at solobase.dev/pricing/', warning: null };
      }
      if (sub.status === 'past_due') {
        if (sub.grace_period_end && new Date(sub.grace_period_end) < new Date()) {
          return { error: 'Payment overdue. Service suspended. Please update payment at cloud.solobase.dev/blocks/dashboard/', warning: null };
        }
        // Within grace period — allow but warn
        const daysLeft = sub.grace_period_end
          ? Math.max(0, Math.ceil((new Date(sub.grace_period_end).getTime() - Date.now()) / 86400000))
          : 0;
        ok.warning = `Payment failed. ${daysLeft} days remaining before service suspension. Update payment at cloud.solobase.dev/blocks/dashboard/`;
      }
    }
  }

  // First, read current state to check limits BEFORE incrementing.
  // This prevents DoS where an attacker inflates the counter by spamming
  // requests that would be rejected anyway.
  const row = await db.prepare(
    `SELECT requests, addon_requests FROM project_usage WHERE project_id = ? AND month = ?`
  ).bind(project.id, month).first<{ requests: number; addon_requests: number }>();

  const currentRequests = row?.requests ?? 0;
  const addonRequests = row?.addon_requests ?? 0;
  const maxRequests = limits.maxRequestsPerMonth + addonRequests;

  // Check limit BEFORE incrementing
  if (currentRequests >= maxRequests) {
    return {
      error: `Plan limit exceeded: ${currentRequests.toLocaleString()} / ${maxRequests.toLocaleString()} requests this month. Upgrade your plan or add more requests.`,
      warning: null,
    };
  }

  // Under limit — upsert and increment
  await db.prepare(
    `INSERT INTO project_usage (id, project_id, month, requests, r2_bytes, addon_requests, addon_r2_bytes, addon_d1_bytes)
     VALUES (?, ?, ?, 1, 0, 0, 0, 0)
     ON CONFLICT (project_id, month) DO UPDATE SET requests = requests + 1`
  ).bind(
    `${project.id}:${month}`, project.id, month
  ).run();

  // Warn at 80% usage
  const newCount = currentRequests + 1;
  const usagePct = newCount / maxRequests;
  if (usagePct >= 0.8 && !ok.warning) {
    const pct = Math.round(usagePct * 100);
    ok.warning = `${pct}% of monthly API requests used (${newCount.toLocaleString()} / ${maxRequests.toLocaleString()})`;
  }

  return ok;
}

/**
 * Check R2 storage usage before allowing an upload.
 */
export async function checkStorageLimit(
  db: D1Database,
  project: ProjectConfig,
  additionalBytes: number,
): Promise<string | null> {
  const limits = getPlanLimits(project.plan);
  const month = currentMonth();

  // Get current R2 usage from storage_objects table
  const row = await db.prepare(
    `SELECT COALESCE(SUM(size), 0) as total_bytes FROM storage_objects`
  ).first<{ total_bytes: number }>();

  const usageRow = await db.prepare(
    `SELECT addon_r2_bytes FROM project_usage WHERE project_id = ? AND month = ?`
  ).bind(project.id, month).first<{ addon_r2_bytes: number }>();

  const currentBytes = row?.total_bytes ?? 0;
  const addonBytes = usageRow?.addon_r2_bytes ?? 0;
  const maxBytes = limits.maxR2StorageBytes + addonBytes;

  if (currentBytes + additionalBytes > maxBytes) {
    const usedMB = Math.round(currentBytes / 1024 / 1024);
    const maxMB = Math.round(maxBytes / 1024 / 1024);
    return `Storage limit exceeded: ${usedMB}MB / ${maxMB}MB. Upgrade your plan or add more storage.`;
  }

  return null;
}

/**
 * Get current usage summary for a project (for dashboard display).
 */
export async function getUsageSummary(
  db: D1Database,
  project: ProjectConfig,
): Promise<{
  plan: string;
  month: string;
  requests: { used: number; limit: number };
  r2Storage: { usedBytes: number; limitBytes: number };
  d1Storage: { usedBytes: number; limitBytes: number };
}> {
  const limits = getPlanLimits(project.plan);
  const month = currentMonth();

  const usage = await db.prepare(
    `SELECT requests, addon_requests, addon_r2_bytes, addon_d1_bytes
     FROM project_usage WHERE project_id = ? AND month = ?`
  ).bind(project.id, month).first<{
    requests: number;
    addon_requests: number;
    addon_r2_bytes: number;
    addon_d1_bytes: number;
  }>();

  const r2 = await db.prepare(
    `SELECT COALESCE(SUM(size), 0) as total FROM storage_objects`
  ).first<{ total: number }>();

  return {
    plan: project.plan,
    month,
    requests: {
      used: usage?.requests ?? 0,
      limit: limits.maxRequestsPerMonth + (usage?.addon_requests ?? 0),
    },
    r2Storage: {
      usedBytes: r2?.total ?? 0,
      limitBytes: limits.maxR2StorageBytes + (usage?.addon_r2_bytes ?? 0),
    },
    d1Storage: {
      usedBytes: 0, // D1 doesn't expose per-database size easily
      limitBytes: limits.maxD1StorageBytes + (usage?.addon_d1_bytes ?? 0),
    },
  };
}

function currentMonth(): string {
  const d = new Date();
  return `${d.getUTCFullYear()}-${String(d.getUTCMonth() + 1).padStart(2, '0')}`;
}

File: crates/solobase-cloudflare/src/usage.rs
//! Account-level usage tracking and plan limit enforcement.
//!
//! Usage is tracked per-project per-month in `project_usage` (for breakdown visibility),
//! but limits are enforced at the account level by summing across all user's projects.
//! Add-on allowances live on the `subscriptions` table (account-level).
//!
//! The check (read) is blocking — needed before dispatching.
//! The increment (write) is non-blocking — runs via waitUntil after response.

use worker::*;

use crate::project::{get_plan_limits, ProjectConfig};

/// Result of a usage check.
pub struct UsageCheckResult {
    /// If set, the request should be rejected with 429.
    pub error: Option<String>,
    /// If set, add as X-Solobase-Warning header (request still succeeds).
    pub warning: Option<String>,
}

impl UsageCheckResult {
    fn ok() -> Self {
        Self { error: None, warning: None }
    }
}

/// Account-level addons read from the subscriptions table.
struct AccountAddons {
    requests: u64,
    r2_bytes: u64,
    d1_bytes: u64,
}

/// Check if the project's owner is within account-level plan limits.
///
/// Sums usage across ALL of the owner's projects for the current month,
/// then checks against plan limits + account-level addons.
pub async fn check_usage(
    db: &D1Database,
    project: &ProjectConfig,
) -> UsageCheckResult {
    let limits = get_plan_limits(&project.plan);
    let month = current_month();
    let mut result = UsageCheckResult::ok();

    let owner_id = match &project.owner_user_id {
        Some(id) if !id.is_empty() => id.clone(),
        _ => return result, // No owner = no limits (platform project)
    };

    // Check subscription status
    match check_subscription_status(db, &owner_id).await {
        SubStatus::Active => {}
        SubStatus::PastDue(warning) => {
            result.warning = Some(warning);
        }
        SubStatus::Inactive(err) => {
            return UsageCheckResult { error: Some(err), warning: None };
        }
    }

    // Free tier: block all requests (no resources)
    if limits.max_requests_per_month == 0 {
        let addons = get_account_addons(db, &owner_id).await;
        if addons.requests == 0 {
            return UsageCheckResult {
                error: Some("No active subscription. Subscribe to a plan to access your project.".into()),
                warning: None,
            };
        }
    }

    // Sum usage across ALL owner's projects for this month
    let (total_requests, total_r2, total_d1) = get_account_usage(db, &owner_id, &month).await;

    // Get account-level addons from subscriptions table
    let addons = get_account_addons(db, &owner_id).await;

    // --- Check request limit ---
    let max_requests = limits.max_requests_per_month.saturating_add(addons.requests);

    if max_requests > 0 && total_requests >= max_requests {
        return UsageCheckResult {
            error: Some(format!(
                "Plan limit exceeded: {} / {} requests this month. Upgrade your plan or purchase add-ons.",
                total_requests, max_requests
            )),
            warning: None,
        };
    }

    // --- Check R2 storage limit ---
    let max_r2 = limits.max_r2_storage_bytes.saturating_add(addons.r2_bytes);

    if max_r2 > 0 && total_r2 >= max_r2 {
        return UsageCheckResult {
            error: Some(format!(
                "Object storage limit exceeded: {} / {} bytes. Upgrade your plan or purchase add-ons.",
                total_r2, max_r2
            )),
            warning: None,
        };
    }

    // --- Check D1 storage limit ---
    let max_d1 = limits.max_d1_storage_bytes.saturating_add(addons.d1_bytes);

    if max_d1 > 0 && total_d1 >= max_d1 {
        return UsageCheckResult {
            error: Some(format!(
                "Database storage limit exceeded: {} / {} bytes. Upgrade your plan or purchase add-ons.",
                total_d1, max_d1
            )),
            warning: None,
        };
    }

    // --- Warnings at 80% usage ---
    if max_requests > 0 {
        let pct = ((total_requests + 1) as f64 / max_requests as f64 * 100.0) as u64;
        if pct >= 80 && result.warning.is_none() {
            result.warning = Some(format!(
                "{}% of monthly API requests used ({} / {})",
                pct, total_requests + 1, max_requests
            ));
        }
    }

    if max_r2 > 0 && total_r2 > 0 {
        let pct = (total_r2 as f64 / max_r2 as f64 * 100.0) as u64;
        if pct >= 80 && result.warning.is_none() {
            result.warning = Some(format!(
                "{}% of object storage used ({} / {} bytes)",
                pct, total_r2, max_r2
            ));
        }
    }

    if max_d1 > 0 && total_d1 > 0 {
        let pct = (total_d1 as f64 / max_d1 as f64 * 100.0) as u64;
        if pct >= 80 && result.warning.is_none() {
            result.warning = Some(format!(
                "{}% of database storage used ({} / {} bytes)",
                pct, total_d1, max_d1
            ));
        }
    }

    result
}

/// Increment the per-project usage counter (non-blocking, called via waitUntil).
/// Tracking stays per-project for breakdown visibility.
pub async fn increment_usage(db: &D1Database, project_id: &str) {
    let month = current_month();
    let usage_id = format!("{}:{}", project_id, month);
    if let Ok(stmt) = db.prepare(
        "INSERT INTO project_usage (id, project_id, month, requests, r2_bytes, d1_bytes) \
         VALUES (?1, ?2, ?3, 1, 0, 0) \
         ON CONFLICT (project_id, month) DO UPDATE SET requests = requests + 1"
    ).bind(&[usage_id.into(), project_id.into(), month.into()]) {
        let _ = stmt.run().await;
    }
}

// ---------------------------------------------------------------------------
// Account-level usage aggregation
// ---------------------------------------------------------------------------

/// Sum requests, r2_bytes, and d1_bytes across all of a user's projects for a month.
async fn get_account_usage(db: &D1Database, owner_id: &str, month: &str) -> (u64, u64, u64) {
    let query = "\
        SELECT COALESCE(SUM(pu.requests), 0) as total_requests, \
               COALESCE(SUM(pu.r2_bytes), 0) as total_r2, \
               COALESCE(SUM(COALESCE(pu.d1_bytes, 0)), 0) as total_d1 \
        FROM project_usage pu \
        JOIN projects p ON p.id = pu.project_id \
        WHERE p.owner_user_id = ?1 AND pu.month = ?2";

    match db.prepare(query).bind(&[owner_id.into(), month.into()]) {
        Ok(stmt) => match stmt.first::<serde_json::Value>(None).await {
            Ok(Some(row)) => (
                row.get("total_requests").and_then(|v| v.as_u64()).unwrap_or(0),
                row.get("total_r2").and_then(|v| v.as_u64()).unwrap_or(0),
                row.get("total_d1").and_then(|v| v.as_u64()).unwrap_or(0),
            ),
            _ => (0, 0, 0),
        },
        Err(_) => (0, 0, 0),
    }
}

/// Read account-level addon allowances from the subscriptions table.
async fn get_account_addons(db: &D1Database, owner_id: &str) -> AccountAddons {
    let query = "SELECT \
        COALESCE(addon_requests, 0) as addon_requests, \
        COALESCE(addon_r2_bytes, 0) as addon_r2_bytes, \
        COALESCE(addon_d1_bytes, 0) as addon_d1_bytes \
        FROM subscriptions WHERE user_id = ?1 AND status = 'active' LIMIT 1";

    match db.prepare(query).bind(&[owner_id.into()]) {
        Ok(stmt) => match stmt.first::<serde_json::Value>(None).await {
            Ok(Some(row)) => AccountAddons {
                requests: row.get("addon_requests").and_then(|v| v.as_u64()).unwrap_or(0),
                r2_bytes: row.get("addon_r2_bytes").and_then(|v| v.as_u64()).unwrap_or(0),
                d1_bytes: row.get("addon_d1_bytes").and_then(|v| v.as_u64()).unwrap_or(0),
            },
            _ => AccountAddons { requests: 0, r2_bytes: 0, d1_bytes: 0 },
        },
        Err(_) => AccountAddons { requests: 0, r2_bytes: 0, d1_bytes: 0 },
    }
}

// ---------------------------------------------------------------------------
// Subscription status check
// ---------------------------------------------------------------------------

enum SubStatus {
    Active,
    PastDue(String),
    Inactive(String),
}

async fn check_subscription_status(db: &D1Database, user_id: &str) -> SubStatus {
    let row = match db
        .prepare("SELECT status, grace_period_end FROM subscriptions WHERE user_id = ?1 LIMIT 1")
        .bind(&[user_id.into()])
    {
        Ok(stmt) => match stmt.first::<serde_json::Value>(None).await {
            Ok(Some(row)) => row,
            _ => return SubStatus::Active, // no subscription = free tier
        },
        Err(_) => return SubStatus::Active,
    };

    let status = row.get("status").and_then(|v| v.as_str().map(String::from)).unwrap_or_default();
    let grace_end = row.get("grace_period_end").and_then(|v| v.as_str().map(String::from));

    match status.as_str() {
        "cancelled" | "suspended" => {
            SubStatus::Inactive("Subscription inactive. Please resubscribe.".into())
        }
        "past_due" => {
            if let Some(ref end) = grace_end {
                if let Ok(end_time) = chrono::DateTime::parse_from_rfc3339(end) {
                    let now = chrono::Utc::now();
                    let end_utc = end_time.with_timezone(&chrono::Utc);
                    if now > end_utc {
                        return SubStatus::Inactive(
                            "Payment overdue. Service suspended. Please update payment.".into(),
                        );
                    }
                    let days_left = (end_utc - now).num_days().max(0);
                    return SubStatus::PastDue(format!(
                        "Payment failed. {} days remaining before suspension.",
                        days_left
                    ));
                }
            }
            SubStatus::PastDue("Payment failed. Please update your payment method.".into())
        }
        _ => SubStatus::Active,
    }
}

fn current_month() -> String {
    chrono::Utc::now().format("%Y-%m").to_string()
}
